Updates for PHP8

colorful-tones opened the pull request:
- Update valid license from `GPL-2.0+` to `GPL-2.0-or-later`
- Update dependencies to latest
- Allow `composer/installers` and `boxuk/wp-muplugin-loader`

## Testing instructions

On PHP 8.2+

1. Check for valid `composer.json` with `composer validate`
2. Lint all files with PHP 8 `find . -type f -name "*.php" ! -path "*/vendor/*" -exec php -l {} \; | grep -v "No syntax errors" || true`

Drop this in your `composer.json`, run `composer install` and it should put `wp-content/mu-plugins/wp-testimonials/` with `wp-content/mu-plugins/mu-loader.php` (generated by `boxuk/wp-muplugin-loader`)

```json
{
    "name": "fun-composer/project",
    "type": "project",
    "repositories": [
        {
            "type": "composer",
            "url": "https://wpackagist.org"
        }
    ],
    "require": {
        "deliciousbrains/wp-testimonials": "dev-php8-updates as dev-master"
    },
    "config": {
        "allow-plugins": {
            "boxuk/wp-muplugin-loader": true,
            "composer/installers": true
        }
    },
    "extra": {
        "installer-paths": {
          "wp-content/mu-plugins/{$name}/": ["type:wordpress-muplugin"],
          "wp-content/plugins/{$name}/": ["type:wordpress-plugin"],
          "wp-content/themes/{$name}/": ["type:wordpress-theme"]
        }
    }
}
```

colorful-tones committed:
Updates for PHP8

polevaultweb commented:
@colorful-tones Does `composer/installers` completely cover the use case of the boxuk package? That recommends using https://packagist.org/packages/roots/bedrock-autoloader instead

colorful-tones commented:
> @colorful-tones Does composer/installers completely cover the use case of the boxuk package? That recommends using https://packagist.org/packages/roots/bedrock-autoloader instead

@polevaultweb - We were only utilizing and relying on `boxuk/wp-muplugin-loader` package's ability to place `wp-content/mu-plugins/wp-testimonials` 👈 in this location with this in the `composer.json`: `"type": "wordpress-muplugin",`. This is now native to `composer/installers`([link to documentation](https://github.com/composer/installers?tab=readme-ov-file#current-supported-package-types:~:text=wolfcms%2Dplugin-,WordPress,-wordpress%2Dplugin%0Awordpress)).

As it stands today, if you were to spin up a WordPress site, `composer init` and add `deliciousbrains/wp-testimonials` then it would do the following (note the reliance on `deliciousbrains/wp-post-types`):

```bash
.
├── composer.json
└── wp-content/
    ├── mu-plugins/
    │   └── wp-testimonials
    └── vendor/
        └── wp-post-types
```

With this PR we obtain the same setup:

```
.
├── composer.json
└── wp-content/
    ├── mu-plugins/
    │   └── wp-testimonials
    └── vendor/
        └── wp-post-types
```

Both `boxuk/wp-muplugin-loader` and `roots/bedrock-autoloader` offer more features besides placing `"type": "wordpress-muplugin",` in `wp-content/mu-plugins`, but we're not utilizing any of them in this repo, and the configuration of our internal sites that rely on these DeliciousBrains Composer dependencies is doing nothing custom either. So, reliance on `boxuk/wp-muplugin-loader` and `roots/bedrock-autoloader` offer no benefit in this repo.

colorful-tones commented:
@polevaultweb  - The fact that we have the `"type": "wordpress-muplugin"` explicitly set in the Composer package is problematic. Ideally, this is a plugin, and with `composer/installers` a developer can require this package as a dependency and override where it is installed (should they prefer it be a mu-plugin) with `composer/installers` by setting this in their `composer.json`, and having this be `"type": "wordpress-plugin"` instead would be a better default behavior.

Here might be an example `composer.json` in a developer's project should they want to use WP Testimonials as a `mu-plugin`, but instead it had `"type": "wordpress-plugin"` in this repo's `composer.json`

```json
{
  "extra": {
    "installer-paths": {
      "wp-content/plugins/{$name}/": [
        "type:wordpress-plugin"
      ],
      "wp-content/mu-plugins/{$name}/": [
        "deliciousbrains/wp-testimonials"
      ]
    }
  }
}
```

polevaultweb commented:
> We were only utilizing and relying on boxuk/wp-muplugin-loader package's ability to place wp-content/mu-plugins/wp-testimonials 👈 in this location with this in the composer.json

@colorful-tones There's more to it than that. 

MU Plugins are single files. `composer/installers ` and `"type": "wordpress-muplugin"` ensure that package gets installed in the `wp-content/mu-plugins` directory rather than in vendor.

However, we have our packages like `wp-post-types` as directories, not just single PHP files. that live in mu-plugins, but WordPress doesn't load directories by default. That is what boxuk was doing for us.

I would assume that if you run the site with these packages on these branches installed, the site would fatal or have an error.

The best way to test this is to have each package checked out locally and on this branch, then on your local install of the site run:

```
composer config repositories.0 path {/path/to/local/package}
composer require deliciousbrains/{package}:dev-master
```

eg.
```
composer config repositories.0 path /sites/deliciousbrains/wp-testimonials
composer require deliciousbrains/wp-testimonials:dev-master
````

You can revert composer.json changes and run `composer install` to get back to normal after testing.

Happy to chat more or screenshare if you get stuck 👍

colorful-tones committed:
Add boxuk/wp-muplugins-loader back as 2.0

colorful-tones committed:
Avoid generating wp-content/mu-plugins/mu-require.php in this package

colorful-tones commented:
Ok, I've added `boxuk/wp-muplugin-loader` back as a dependency and updated its version from `^1.0` to `^2.0`. I investigated `roots/bedrock-loader` and it only generates an mu-plugins auto-loader if you use the Bedrocks directory structure: `web/app/mu-plugins/`. 

In my experience having a Composer package generate the entrypoint for loading mu-plugins is a bit of an anti-pattern. However, it was already an experience that users of this WP Testimonials package have come to expect. Therefore, we need to ensure backwards-compatiblity for anyone that has this public package (from Packagist).

Here is a basic `composer.json` of how you would now use this in your project (note: I'm pointing version to this branch for testing):

```
{
    "name": "fun-composer/project",
    "type": "project",
    "repositories": [
        {
            "type": "composer",
            "url": "https://wpackagist.org"
        }
    ],
    "require": {
        "deliciousbrains/wp-testimonials": "dev-php8-updates as dev-master"
    },
    "config": {
        "allow-plugins": {
            "boxuk/wp-muplugin-loader": true,
            "composer/installers": true
        }
    },
    "extra": {
        "installer-paths": {
          "wp-content/mu-plugins/{$name}/": ["type:wordpress-muplugin"],
          "wp-content/plugins/{$name}/": ["type:wordpress-plugin"],
          "wp-content/themes/{$name}/": ["type:wordpress-theme"]
        }
    }
}
```

@polevaultweb let me know if this works?

polevaultweb commented:
@colorful-tones works for me, thanks 👍
